### PR TITLE
[Perf][Diffusion] Keep Helios modulation compact across tokens

### PR DESCRIPTION
**Pull request by scestola-h**
## Purpose

Before its transformer blocks, Helios expands the per-block modulation to every token:
`timestep_proj` becomes `[B, S, 6, D]` and `temb` becomes `[B, S, D]`. Yet each batch
item carries at most two distinct conditions: the current timestep and, when
`zero_history_timestep` is set, the zero-timestep row used for history tokens. With
`D = 5120` and 40 blocks, every block then broadcasts sequence-length tensors for two
scale/shift pairs and two gates.

This PR keeps the modulation compact. It stores `[R, 6, D]` rows, where `R = B`, or
`2B` with a history condition, plus one row index per flattened token. The block
scale/shift and gates then go through the shared `indexed_scale_shift_` and
`indexed_gate` kernels in `vllm_omni/diffusion/layers/indexed_modulation.py`, which
MiniMax-H3 already uses. The arithmetic matches the existing Helios block: FP32
accumulation, then a cast back to the hidden dtype. It is bit-exact in the measured
Ascend eager runs; on CUDA, Triton may contract multiply-adds, so the CUDA reference
test uses `atol=rtol=1e-5`. The output normalization reads only the current condition,
so it broadcasts a single `[B, 1, D]` row.

Changes:

- `vllm_omni/diffusion/models/helios/helios_transformer.py`:
  - `_make_compact_modulation` builds the rows and the token indices.
  - Blocks take `modulation, modulation_indices`.
  - The sequence-length expansion, the unused `temb_t0`, and a slice in
    `HeliosOutputNorm` that is now a no-op are removed.
- `tests/diffusion/models/helios/test_helios_modulation.py`:
  - A CPU test compares the production rows and indices with an independent per-token
    expansion. It covers no history, a history condition, and history tokens without
    the condition, in FP32 and BF16, at hidden size 5120 and batch 2.
  - A3 and CUDA tests check both kernels at Helios width against eager references. The
    CUDA cases preserve production's strided `unbind(1)` row layout and include a
    standalone dynamic-shape, full-graph compile check of a block-like helper.
- `.buildkite/npu/test-npu-ready.yml`: runs the new test in the A3 layer-test step,
  next to the MiniMax-H3 modulation test.

## Test Plan

Focused tests on Ascend A3:

```bash
pytest -v tests/diffusion/models/helios/test_helios_modulation.py \
  tests/diffusion/models/helios/test_helios_usp_shard.py \
  tests/diffusion/models/minimax_h3/test_minimax_h3_modulation.py
```

CUDA correctness and a standalone compile check on one NVIDIA L4 (the same GPU class
used by the CUDA model job):

```bash
pytest -v tests/diffusion/models/helios/test_helios_modulation.py \
  -m 'core_model and cuda' --run-level core_model
```

End-to-end A/B, one fresh process per run on one die, with arms alternated:

```bash
python examples/offline_inference/text_to_video/text_to_video.py \
  --model /path/to/Helios-Distilled \
  --prompt "A dynamic time-lapse video showing the rapidly moving scenery from the window of a speeding train." \
  --num-frames {33,99} --num-inference-steps 50 --guidance-scale 1.0 \
  --height 384 --width 640 --seed 1101 --fps 16 \
  --enable-diffusion-pipeline-profiler \
  --extra-body '{"is_enable_stage2": true, "pyramid_num_inference_steps_list": [2, 2, 2], "is_amplify_first_chunk": true}' \
  --enforce-eager --vae-use-slicing --vae-use-tiling
```

**vLLM Version:** `0.29.0` (`quay.io/atlas-ci/vllm-ascend:v0.29.0-a3`, vllm-ascend
`0.19.1rc2.dev2045+g12be3d34a`)

**vLLM-Omni Commit:** both measured A/B arms were built from `e6cebf85`. That commit
predates an open fix for #6057, without which the stage-2 pyramid path used in this A/B
fails on Ascend. Both arms therefore carried the same local block-noise fix, which is not
part of this PR. The arms differ only in this PR's files. Before submission, the branch
was rebased onto `eb1abbad`; the intervening upstream commits do not touch Helios, the
shared indexed-modulation kernels, the new test path or the edited NPU CI file.

## Test Result

Environment: Ascend 910_9362, CANN 9.1.0, torch 2.10.0, torch_npu 2.10.0.post4,
triton-ascend 3.2.2, one die. Runs used eager mode. The NPU platform reports
`supports_torch_inductor() == False`, so eager is also Ascend's default path.

| | base | this PR |
|---|---:|---:|
| 33 frames: generation time (mean of 2, after 1 warmup per arm) | 23.93 s | 19.81 s (**1.21×**) |
| 33 frames: excluding VAE decode | 20.57 s | 16.45 s (**1.25×**) |
| 33 frames: peak device memory | 53,169 MB | 46,026 MB (**−7.1 GB**) |
| 99 frames: generation time (base 1 run, PR 2 runs) | 51.12 s | 42.83 s (**1.19×**) |
| 99 frames: excluding VAE decode | 41.04 s | 32.77 s (**1.25×**) |
| 99 frames: peak device memory | 53,174 MB | 46,436 MB (**−6.7 GB**) |

- **Byte-identical output.** Base and PR produce byte-identical MP4s and identically
  decoded frames. The SHA-256 prefix is `561defdd1584db6d` at 33 frames and
  `4a20febb6f64f739` at 99 frames. Every run in both arms reproduced its hash: 3 base and
  3 PR runs at 33 frames, 1 base and 2 PR runs at 99 frames.
- **History conditions covered.** The 33-frame runs use the first-chunk path, which
  already carries zero-initialised history with the zero-timestep condition. The 99-frame
  runs add two chunks with real history.
- **Timing spread and scope.** Repeated timings differed by at most 0.06 s. Generation
  time excludes engine start-up and weight loading. Peak memory was sampled from
  `npu-smi` every 0.5 s.
- **Teardown abort.** Every run, in both arms, completed generation and saved its output,
  and each process then hit a known exit-time allocator abort on this stack
  (`corrupted size vs. prev_size`, rc=134). The pytest runs did the same after reporting
  their results.
- **Focused A3 tests:** 30 passed, 1 skipped. The 30 are 8 new Helios tests, 15 Helios USP
  shard tests and 7 MiniMax-H3 modulation tests. Three local block-noise regression tests
  also passed. The two CUDA cases were added afterwards; on A3 they would be skipped.
- **CUDA tests:** on one NVIDIA L4 with driver 595.71.05, PyTorch 2.13.0+cu130,
  Triton 3.7.1 and vLLM 0.29.0, the CI-style CUDA selection reported 2 passed and 8
  deselected in 5.22 s. The complete new module reported 8 passed and 2 skipped in
  5.48 s. The raw FP32 kernel comparisons use `atol=rtol=1e-5`; the compiled block-like
  helper returns BF16 and uses `assert_close`'s dtype-aware BF16 tolerance.
- **`test_helios_usp_2rank.py`** fails the same way before and after this PR on this
  stack. `HeliosTransformer3DModel` construction reaches vLLM `CustomOp` via
  `Conv3dLayer`, and the test never sets a vLLM config. This test is marked `full_model`
  and is not selected by the ready pipeline. Locally, I wrapped model construction in
  `set_current_vllm_config(VllmConfig())` and worked around an Ascend-specific
  process-exit abort. With those changes, ws=2 matched ws=1 exactly in both arms, and
  base and PR outputs were bit-identical at ws=1 and ws=2. Neither change is part of this
  PR.
- Diff-scoped pre-commit hooks passed after the CUDA cases were added: ruff, ruff-format,
  typos, mypy-3.10, check-mark, SPDX, forbidden imports, `torch.cuda` calls, YAML and
  Buildkite validation.

The CUDA cases exercise both raw kernels at Helios width against eager references and a
block-like helper at two sequence lengths through
`torch.compile(dynamic=True, fullgraph=True)`. They do not exercise Helios's actual
regional-compile boundary (`dynamic=True`, per block), and Cache-DiT was not
smoke-tested. Full-model CUDA output and performance remain untested. The byte-identity
and performance results above cover Ascend only.

AI assistance: AI coding assistants were used to review the implementation, run the
Ascend and CUDA validation, and draft this description. The submitted diff and technical
claims were reviewed by the contributor; the validation performed and the checks not run
are reported above.

### CODE REVIEW

**Comment by scestola-h** [comment]
Author self-review:

- Reviewed the compact row construction and token-index mapping through the transformer blocks and output normalization, including no-history and history-conditioned paths. The mapping preserves the previous per-token conditions while removing the sequence-wide modulation expansion.
- Checked the dtype and numerical boundaries: the indexed kernels accumulate in FP32 and cast back at the same boundary as the prior implementation. The CUDA reference tolerance accounts for possible fused multiply-add contraction.
- Reviewed the new coverage: an independent CPU mapping reference, A3 and CUDA kernel tests at Helios width, and a CUDA dynamic-shape/full-graph block-like helper at two sequence lengths.
- Validation completed: focused A3 tests passed; the measured A3 end-to-end outputs were byte-identical to base while reducing peak device memory by 6.7–7.1 GB and improving generation time by 1.19–1.21x. On an NVIDIA L4, the complete new test module reported 8 passed and 2 skipped.
- Remaining limitations are called out in the PR body: the actual full-model CUDA regional-compile path, CUDA end-to-end performance, and Cache-DiT were not exercised.

I found no blocking issues in the submitted diff.

**Comment by Dong1017** [comment]
Reviewed `4c18b864`. Change surface is minimal: the win comes from reusing the existing shared `indexed_modulation` kernels rather than adding new ones; production changes are confined to `helios_transformer.py` (compact rows + token indices, block signature, output-norm broadcast, dead-code removal), and the test file's size is justified by the independent per-token reference plus its CPU/CUDA/A3 variants.

Verified the equivalence: the row/index mapping reproduces the old `[t0; current]` expansion exactly — the old `indices_hidden_states is not None` conditions were always true after the forward's default, so the new `history_context_length > 0` guard is equivalent; FP32 accumulate-then-cast and the bf16 branch upcast match the old expression; the in-place affine targets the fresh FP32LayerNorm output; production's `unbind(1)` strided rows are asserted in the CUDA test; the model forward is the only caller of the changed block signature. All three test variants are wired into matching Buildkite lanes.

The Ascend results (byte-identical E2E, −6.7–7.1 GB, 1.19–1.21×) are author evidence, consistent with the bit-exact design. Still owed before full qualification (author-declared): full-model CUDA E2E, regional compile, Cache-DiT, CUDA SP. Static review of this head, no GPU run needed.
